Project: V-A-M-P-I-R-E-S

CringeCursed opened the pull request:
## Short description
port from: https://paradisestation.org/wiki/index.php/Vampire
![drasticlyassvampire](https://github.com/user-attachments/assets/b6e0eb86-5588-45d1-a80c-1c6e4d69fbfd)

## Why we need to add this

## Media (Video/Screenshots)
test it locally im too lazy to show all of this
(ill add media later)

## Checks

- [x] I do not require assistance to complete the PR.
- [x] Before posting/requesting review of a PR, I have verified that the changes work.
- [ ] I have added screenshots/videos of the changes, or this PR does not change in-game mechanics.
- [x] I affirm that my changes are licensed under the [Starlight Fork License](https://github.com/ss14Starlight/space-station-14/blob/Starlight/LICENSE-Starlight.TXT) and grant permission for use in this repository under its conditions.

**Changelog**

:cl: CringeCursed, Rinary, DocMoth, YourSly
- tweak: New vampires

CringeCursed committed:
init

CringeCursed committed:
basesystemsstartup and 2 base abilities

CringeCursed committed:
ifuckedlaw

Rinary1 committed:
some refactors

Rinary1 committed:
fixes of fixes

Rinary1 committed:
Merge branch 'starlight-dev' of github.com:ss14Starlight/space-station-14 into Vumpires-reworked

# Conflicts:
#	Content.Server/GameTicking/Rules/VampireRuleSystem.cs
#	Content.Server/Vampire/VampireSystem.Abilities.cs
#	Content.Server/Vampire/VampireSystem.cs
#	Resources/Prototypes/Entities/Mobs/Species/base.yml
#	Resources/Prototypes/_StarLight/tags.yml

Rinary1 committed:
misc fixes

Rinary1 committed:
huge refactors

Rinary1 committed:
Merge branch 'starlight-dev' of github.com:ss14Starlight/space-station-14 into Vumpires-reworked

Rinary1 committed:
refactors

Rinary1 committed:
more refactors

Rinary1 committed:
Merge branch 'starlight-dev' of github.com:ss14Starlight/space-station-14 into Vumpires-reworked

Rinary1 committed:
hemomancer refactors

Rinary1 committed:
more n more refactors

Rinary1 committed:
Merge branch 'starlight-dev' of github.com:ss14Starlight/space-station-14 into Vumpires-reworked

Rinary1 committed:
refactors

Rinary1 committed:
remove unused type

Rinary1 committed:
fix umbrae

Rinary1 committed:
fix issues in umbrae(move all to locale/resources)

Rinary1 committed:
fix all hemomancer resources issues

Rinary1 committed:
more fixes

CringeCursed committed:
Merge remote-tracking branch 'upstream/vumpires-reworked' into vamps-rework-continuew

CringeCursed committed:
Dantaleon Class + Handling Space("Sun") exposure

CringeCursed committed:
Gargantua(Shitmaded) gotta hardly redo
Ftl(shitmaded) that either
Need Sprites still

CringeCursed committed:
blood swell & blood rush refactor
Alerts to status bar as well for abilities(not all just swell and rush)
bunch of refactors for almost every abilities

CringeCursed committed:
bunch of refactors, clean-up some warnings, sounds etc.

CringeCursed committed:
New Sprite Effect for shadow boxing

CringeCursed committed:
A lot of refactors and big ass cleanp Vampire Systems

CringeCursed committed:
collective mind for thralls and fix: class ui

CringeCursed committed:
Last Ability for Hemomancer
Full Refactor of Shadow Anchor Ability
More roundend Stats
GamePreset changes
Fixes BUI for MobSlime
Cleanp

CringeCursed committed:
cleanup

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
- DidYouPrayThatDay?
Fix: HysteriaVision offset
Tweak: Refactored tenderils hemomancer ability
add: Guides to vampire
fix: Vampire Rule and Antag selection
fix: Rejuvenate ability,its upgrade and Rally thrals to actually remove knockdown
add: vulnerability to holy places and holy water to vampire
and few more things lmao

CringeCursed committed:
copyright update

CringeCursed committed:
remove obsolete icons

CringeCursed committed:
fixes

CringeCursed committed:
more fixes

CringeCursed committed:
Ifuckedaw

CringeCursed committed:
fix desc in actions

CringeCursed committed:
factionicon

Clone0401 commented:
HOLI MOLI
Actually working vampires!
It’s a Christmas miracle!

PubliclyExecutedPig commented:
As part of testing can you make sure vampires dont burn up on certain stations https://github.com/ss14Starlight/space-station-14/pull/2198 (Test maps like Core and Plasma to see if they do)

CringeCursed commented:
> HOLI MOLI Actually working vampires! It’s a Christmas miracle!

this fucking interation tests :sob:

CringeCursed committed:
Hoppefully no more test fails

ScorchFollower commented:
![pilebunker-slayer-ggst](https://github.com/user-attachments/assets/43dd78da-60de-4ce1-a0ac-da01ec1d3aa2)

We're so back chat, and we can finally pilebunker those filthy secoffs 
<img width="44" height="41" alt="image" src="https://github.com/user-attachments/assets/0b508bf5-0262-401c-a802-1db86476fa87" />

CringeCursed committed:
refactor shadnowsnare ability
add sound for masshysteria

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs:
```diff
@@ -0,0 +1,186 @@
+/// Overlay that renders monster/animal sprites over humanoids  
+/// when the local player has HysteriaVisionComponent.
+/// </summary>
+public sealed class HysteriaVisionOverlay : Robust.Client.Graphics.Overlay
```
```suggestion
public sealed class HysteriaVisionOverlay : Overlay
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs:
```diff
@@ -0,0 +1,186 @@
+using Robust.Shared.Graphics.RSI;
+using Robust.Shared.Random;
+using Robust.Shared.Timing;
+using Robust.Shared.Utility;
```
```suggestion
using Robust.Shared.Utility;
using Robust.Client.Graphics;
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs:
```diff
@@ -0,0 +1,186 @@
+
+        // Check if effect expired
+        if (_timing.CurTime > hysteria.EndTime)
+            return false;
```
```suggestion
        if (player == null 
            || !_entManager.TryGetComponent<HysteriaVisionComponent>(player, out var hysteria) 
            || _timing.CurTime > hysteria.EndTime) // Check if effect expired
            return false;
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs:
```diff
@@ -0,0 +1,186 @@
+            {
+                _disguiseStates[i] = null;
+                continue;
+            }
```
```suggestion
            if (!_resourceCache.TryGetResource<RSIResource>(path, out var rsiResource) 
                || !rsiResource.RSI.TryGetState(sprite.State, out var rsiState))
            {
                _disguiseStates[i] = null;
                continue;
            }
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs:
```diff
@@ -0,0 +1,186 @@
+
+            // Skip entities that are not visible
+            if (!sprite.Visible)
+                continue;
```
```suggestion
            if (xform.MapID != args.MapId // Skip if not on the same map
                || uid == player // Skip self
                || !sprite.Visible) // Skip entities that are not visible
                continue;
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/HysteriaVisionSystem.cs:
```diff
@@ -0,0 +1,91 @@
+        if (_timing.CurTime > hysteria.EndTime)
+        {
+            RemComp<HysteriaVisionComponent>(player.Value);
+        }
```
```suggestion
        if (player == null || !TryComp<HysteriaVisionComponent>(player, out var hysteria))
            return;

        // Remove component if expired
        if (_timing.CurTime > hysteria.EndTime)
            RemComp<HysteriaVisionComponent>(player.Value);
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireBloodBondBeamSystem.cs:
```diff
@@ -0,0 +1,134 @@
+        if (_activeBeamVisuals.TryGetValue(key, out var existingBeam))
+        {
+            QueueDel(existingBeam);
+        }
```
```suggestion
        if (_activeBeamVisuals.TryGetValue(key, out var existingBeam))
            QueueDel(existingBeam);
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireClassBui.cs:
```diff
@@ -0,0 +1,88 @@
+            return;
+
+        if (!EntMan.EntityExists(Owner) || !EntMan.TryGetComponent<MetaDataComponent>(Owner, out _))
+            return;
```
```suggestion
        if (_choiceMade 
            || !EntMan.EntityExists(Owner) || !EntMan.TryGetComponent<MetaDataComponent>(Owner, out _))
            return;
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireDecoySystem.cs:
```diff
@@ -0,0 +1,37 @@
+            return;
+
+        if (!TryComp<SpriteComponent>(ent.Comp.Source.Value, out var sourceSprite))
+            return;
```
```suggestion
        if (!ent.Comp.Source.HasValue 
            || !TryComp<SpriteComponent>(ent.Owner, out var decoySprite) 
            || !TryComp<SpriteComponent>(ent.Comp.Source.Value, out var sourceSprite))
            return;
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireDrainBeamSystem.cs:
Isn't that the same thing as `VampireBloodBondBeamSystem.cs`? If it is, just make one system like:
`VampireBeamVisualsSystem`

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireLocateBui.cs:
```diff
@@ -0,0 +1,152 @@
+
+    protected override void Dispose(bool disposing)
+    {
+        if (disposing && _window != null)
+        {
+            _window.TargetSelected -= OnTargetSelected;
+        }
```
```suggestion
        if (disposing && _window != null)
            _window.TargetSelected -= OnTargetSelected;
```

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireLocateBui.cs:
```diff
@@ -0,0 +1,152 @@
+            HorizontalExpand = true,
+            VerticalExpand = true,
+            SeparationOverride = 2,
+        };
```
isn't easier and better to just add box container with public access to window and then call it like:
`_window.BoxContainerName` to do some stuff, istead of adding it every window creation?

https://github.com/ss14Starlight/space-station-14/blob/af0a3953450b2435ae02fe9aa67087eb973dc55c/Content.Client/_Starlight/Economy/Atm/ATMWindow.xaml#L19

https://github.com/ss14Starlight/space-station-14/blob/af0a3953450b2435ae02fe9aa67087eb973dc55c/Content.Client/_Starlight/Economy/Atm/ATMBui.cs#L154

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireLocateBui.cs:
```diff
@@ -0,0 +1,152 @@
+            };
+
+            button.OnPressed += _ => TargetSelected?.Invoke(captured);
+            _list.AddChild(button);
```
https://github.com/ss14Starlight/space-station-14/pull/2599/changes#r2647156846

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireLocateBui.cs:
```diff
@@ -0,0 +1,152 @@
+
+        if (!any)
+        {
+            _list.AddChild(new Label
```
https://github.com/ss14Starlight/space-station-14/pull/2599/changes#r2647156846

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireLocateBui.cs:
```diff
@@ -0,0 +1,152 @@
+
+    private void Populate()
+    {
+        _list.DisposeAllChildren();
```
https://github.com/ss14Starlight/space-station-14/pull/2599/changes#r2647156846

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireShadowBoxingSystem.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+        _active.Add(new ActivePunch(punch, source, target, _timing.CurTime, _punchLifetime));
+    }
+    // Помогите
```
lol

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireShadowBoxingSystem.cs:
```diff
@@ -0,0 +1,75 @@
+
+        var srcPos = _transform.GetWorldPosition(source);
+
+        var punch = Spawn("VampireShadowBoxingPunch", Transform(source).Coordinates);
```
you can hide this(prototype id) into the event

Rinary1 commented on Content.Client/_Starlight/Antags/Vampires/VampireShadowBoxingSystem.cs:
```diff
@@ -0,0 +1,75 @@
+    private readonly List<ActivePunch> _active = new();
+
+    // How long a punch effect flys for?
+    private static readonly TimeSpan _punchLifetime = TimeSpan.FromSeconds(0.33); 
```
you can hide this into the event

Rinary1 reviewed (changes requested):
Client side review

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+
+    private void OnEnthrallDoAfter(EntityUid uid, VampireComponent comp, ref VampireEnthrallDoAfterEvent args)
+    {
+        if (args.Handled || args.Cancelled || args.Target == null)
+            return;
+
+        if (!TryComp(uid, out DantalionComponent? dantalion))
+            return;
```
```suggestion
        if (args.Handled || args.Cancelled || args.Target == null 
            || !TryComp(uid, out DantalionComponent? dantalion))
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+            return;
+
+        if (!_mind.TryGetMind(master, out var masterMindId, out _))
+            return;
```
```suggestion
        if (!_mind.TryGetMind(thrall, out var thrallMindId, out var thrallMind) 
            || !_mind.TryGetMind(master, out var masterMindId, out _))
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/Classes/DantalionSystem.cs:
```diff
@@ -0,0 +1,798 @@
+        _targetObjectives.SetTarget(objective.Value, masterMindId);
+        _mind.AddObjective(thrallMindId, thrallMind, objective.Value);
+    }
+    private void OnThrallShutdown(EntityUid uid, VampireThrallComponent component, ComponentShutdown args)
+    {
+        if (component.Master is not { } master || !TryComp(master, out DantalionComponent? dantalion))
+            return;
+
+        if (!dantalion.Thralls.Remove(uid))
+            return;
```
```suggestion
    }

    private void OnThrallShutdown(EntityUid uid, VampireThrallComponent component, ComponentShutdown args)
    {
        if (component.Master is not { } master || !TryComp(master, out DantalionComponent? dantalion) 
            || !dantalion.Thralls.Remove(uid))
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+            return false;
+
+        if (!TryComp(uid, out DantalionComponent? dantalionComp))
+            return false;
```
```suggestion
        if (comp.ChosenClass != VampireClassType.Dantalion
            || !TryComp(uid, out DantalionComponent? dantalionComp))
            return false;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+#region Pacify
+    public void OnPacify(EntityUid uid, VampireComponent comp, ref VampirePacifyActionEvent args)
+    {
+        if (args.Handled || !Exists(args.Target))
+            return;
+
+        if (!ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
+            return;
```
```suggestion
        if (args.Handled || !Exists(args.Target) 
            || !ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+#region Subspace Swap
+    private void OnSubspaceSwap(EntityUid uid, VampireComponent comp, ref VampireSubspaceSwapActionEvent args)
+    {
+        if (args.Handled || !Exists(args.Target))
+            return;
+
+        if (!ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
+            return;
```
```suggestion
        if (args.Handled || !Exists(args.Target) 
            || !ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+#region Decoy
+    private void OnDecoy(EntityUid uid, VampireComponent comp, ref VampireDecoyActionEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
+            return;
```
```suggestion
        if (args.Handled || !ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+            return;
+
+        if (args.DamageDelta == null || !args.DamageDelta.AnyPositive())
+            return;
```
```suggestion
        if (component.Detonated || args.DamageDelta == null || !args.DamageDelta.AnyPositive())
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+        if (args.Handled)
+            return;
+
+        if (!ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
+            return;
+
+        if (!TryComp<DantalionComponent>(uid, out var dantalion))
+            return;
```
```suggestion
        if (args.Handled 
            || !ValidateVampireClass(uid, comp, VampireClassType.Dantalion) 
            || !TryComp<DantalionComponent>(uid, out var dantalion))
            return;
```

Rinary1 commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,798 @@
+            return;
+
+        if (!ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
+            return;
```
```suggestion
        if (args.Handled || !ValidateVampireClass(uid, comp, VampireClassType.Dantalion))
            return;
```

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/HysteriaVisionOverlay.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/HysteriaVisionSystem.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/VampireBloodBondBeamSystem.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/VampireClassBui.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
Update Content.Client/_Starlight/Antags/Vampires/VampireDecoySystem.cs

Co-authored-by: Rinary <rinary.super@gmail.com>

CringeCursed committed:
redo Predator Sense to xaml
Unify DrainBeam systems
Move some fields from ShadowBoxing to Event
Add Shared system for BloodSwell ShotAttemptedEvent pop-up
Make Hysteria Overlays thralls for non thralls

CringeCursed committed:
Merge branch 'vamps-rework-continuew' of https://github.com/CringeCursed/space-station-14 into vamps-rework-continuew

CringeCursed committed:
fix

Happyrobot33 commented on .editorconfig:
dont... do this?

CringeCursed replied:
> dont... do this?

i have no clue what ts is
its rinary old ass commit

Rinary1 replied:
> i have no clue what ts is its rinary old ass commit

warnings on unused usings/unnecessary clarifications

CringeCursed committed:
Clean up .editorconfig by removing unused settings

Removed unnecessary diagnostic settings for warnings.

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

Conflee reviewed (approved):
sprites all look good

Penguinwizzard commented on Content.IntegrationTests/Tests/_StarLight/Antags/VampireRuleTest.cs:
Can you put this under a Starlight path?

Penguinwizzard commented on Content.Server/Administration/Systems/AdminVerbSystem.Antags.cs:
```diff
@@ -232,7 +232,7 @@ private void AddAntagVerbs(GetVerbsEvent<Verb> args)
             Text = Loc.GetString("admin-verb-text-make-vampire"),
             Category = VerbCategory.Antag,
-            Icon = new SpriteSpecifier.Rsi(new ResPath("/Textures/Interface/Actions/actions_vampire.rsi"), "unholystrength"),
+            Icon = new SpriteSpecifier.Rsi(new ResPath("/Textures/_Starlight/Vampire/actions_vampire.rsi"), "select_class"),
```
While you're touching this file, can you add Starlight comments around the downstream portion?

Penguinwizzard commented on Content.Server/Antag/AntagSelectionSystem.cs:
```diff
@@ -310,16 +310,16 @@ public void ChooseAntags(Entity<AntagSelectionComponent> ent,
-                if (session != null && HasComp<VampireRuleComponent>(ent))
-                {
-                    var playerEntity = session.AttachedEntity;
+                // if (session != null && HasComp<VampireRuleComponent>(ent))
```
If this no longer makes sense, we should delete it instead of commenting it out (since I assume it's downstream-only?)

CringeCursed committed:
new sprite for blood bond.
prevent beam abilities working through walls
briefing sound

CringeCursed committed:
Merge branch 'vamps-rework-continuew' of https://github.com/CringeCursed/space-station-14 into vamps-rework-continuew

CringeCursed committed:
remove debug values

CringeCursed commented:
<img width="447" height="49" alt="image" src="https://github.com/user-attachments/assets/b9f4fda1-548a-4c64-9801-23274bdf2f9c" />
<img width="755" height="489" alt="image" src="https://github.com/user-attachments/assets/2d3803ad-aafd-4a09-bccd-3c187934bcb4" />

Penguinwizzard commented on Content.Server/GameTicking/Rules/VampireRuleSystem.cs:
```diff
@@ -52,117 +25,99 @@ public override void Initialize()
 
-            _antag.SendBriefing(target, MakeBriefing(target), Color.Yellow, BriefingSound);
-            _role.MindHasRole<VampireRoleComponent>(mindId, out var vampireRole);
+        if (TryComp(target, out MetaDataComponent? meta))
```
```suggestion
        var meta = MetaData(target);
```

Penguinwizzard commented on Content.Server/GameTicking/Rules/VampireRuleSystem.cs:
```diff
@@ -52,117 +25,99 @@ public override void Initialize()
         {
-            var randomEscapeObjective = rule.StealObjectives[random.Next(rule.StealObjectives.Count)];
-            _mind.TryAddObjective(mindId, mind, randomEscapeObjective);
+            var obj = rule.StealObjectives[rng.Next(rule.StealObjectives.Count)];
```
If you change rng to be a RobustRandom dependency reference, you can clean up a few of these to instead do
```suggestion
            var obj = rng.Pick(rule.StealObjectives);
```

Penguinwizzard commented on Content.Server/_Starlight/GameTicking/Rules/VampireRuleSystem.cs:
```diff
@@ -52,117 +25,99 @@ public override void Initialize()
                 continue;
 
-            if (!TryComp<MetaDataComponent>(vamp.Owner, out var metaData))
+            if (!TryComp(vampUid, out MetaDataComponent? meta))
```
```suggestion
            var meta = MetaData(vampUid);
```

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/SanguinePoolSystem.cs:
```diff
@@ -0,0 +1,98 @@
+        if (!_solution.TryGetSolution(uid, puddle.SolutionName, out _, out var solution))
+            return false;
+
+        return solution.ContainsReagent("Blood", null);
```
Does this work for blood spilled by a person? This is looking for blood with specifically no data on it, while person-spilled blood has data (for forensics tracking). You probably want `ContainsPrototype` if you're trying to do a general blood lookup. Also, move from an inline string here to a static readonly `ProtoId<ReagentPrototype>` (which avoids a warning).

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/SanguinePoolSystem.cs:
```diff
@@ -0,0 +1,98 @@
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<SanguinePoolComponent, TransformComponent>();
```
How many of these pools can a vampire have out at once? If we're scanning over them on Update we may want to avoid having too many.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,774 @@
+        args.Handled = true;
+    }
+
+    private void OnDecoyDamaged(EntityUid uid, VampireDecoyComponent component, DamageChangedEvent args)
```
Move to VampireSystem.Decoy.cs as part of changing that to VampireDecoySystem.cs.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Decoy.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Content.Server._Starlight.Antags.Vampires;
+// shitcode
+public sealed partial class VampireSystem
```
This one could probably just be VampireDecoySystem.cs - TriggerDecoyFlash is on an event handler that only looks at VampireDecoyComponent.

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/Components/VampireComponent.cs:
```diff
@@ -0,0 +1,195 @@
+    public int UniqueHumanoidVictims = 0;
+
+    [ViewVariables(VVAccess.ReadOnly), DataField, AutoNetworkedField]
+    public VampireClassType ChosenClass = VampireClassType.None;
```
I'm not a huge fan of this - if we're going to add a component for the chosen class anyway, we could just use the presence of the class component to figure out which class we've chosen, and avoid having two sources of truth.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/Classes/DantalionSystem.cs:
```diff
@@ -0,0 +1,774 @@
+    private void OnEnthrallDoAfter(EntityUid uid, VampireComponent comp, ref VampireEnthrallDoAfterEvent args)
+    {
+        if (args.Handled || args.Cancelled || args.Target == null 
+            || !TryComp(uid, out DantalionComponent? dantalion))
```
This goes for a lot of the methods in this file - if the event only should get triggered for a particular vampire class, we should put the event listener on the class component, and then TryComp the VampireComponent for the few things we need it for.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Dantalion.cs:
```diff
@@ -0,0 +1,774 @@
+using Robust.Shared.Audio;
+using Robust.Shared.Player;
+
+namespace Content.Server._Starlight.Antags.Vampires;
```
The namespace should match the filepath (`Systems` at the end).

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Abilities.cs:
```diff
@@ -0,0 +1,726 @@
+
+namespace Content.Server._Starlight.Antags.Vampires;
+
+public sealed partial class VampireSystem : EntitySystem
```
Why is VampireSystem split across files in `Content.Server/_Starlight/Antags/Vampires/Systems` and `Content.Server/_Starlight/Antags/Vampires`?

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/VampireSystem.Abilities.cs:
```diff
@@ -0,0 +1,726 @@
+
+        var spec = new DamageSpecifier();
+        spec += new DamageSpecifier(damageType, FixedPoint2.New(amount));
+        EntityManager.System<DamageableSystem>().TryChangeDamage(target, spec, true, origin: origin);
```
Should we add a `DamageableSystem` `[Dependency]`?

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Gargantua.cs:
```diff
@@ -0,0 +1,691 @@
+
+    private void OnBloodSwellIncomingDamage(EntityUid uid, GargantuaComponent component, ref BeforeDamageChangedEvent args)
+    {
+        if (!component.BloodSwellActive)
```
I'm on the fence about whether to recommend dropping the "BloodRushActive" and "BloodSwellActive" bools, and replacing them with ActiveBloodRushComponent and ActiveBloodSwellComponent, which get added/removed when the value would otherwise get flipped. It doesn't look like their modifiers get anything else from GargantuaComponent?

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/Classes/GargantuaSystem.cs:
```diff
@@ -0,0 +1,691 @@
+        args.Value *= 0.5f;
+    }
+
+    private void OnStatusEffectApplied(EntityUid effectUid, StatusEffectComponent effect, ref StatusEffectAppliedEvent args)
```
Lightly suggested, since I don't have perf data on the alternatives: This might be more efficient if you changed this line to
```cs
private void OnStatusEffectApplied<T>(EntityUid effectUid, T effect, ref StatusEffectAppliedEvent args)
```
and the listener registration to
```cs
SubscribeLocalEvent<StunnedStatusEffectComponent, StatusEffectAppliedEvent>(OnStatusEffectApplied);
SubscribeLocalEvent<KnockdownStatusEffectComponent, StatusEffectAppliedEvent>(OnStatusEffectApplied);
SubscribeLocalEvent<MovementModStatusEffectComponent, StatusEffectAppliedEvent>(OnStatusEffectApplied);
SubscribeLocalEvent<ForcedSleepingStatusEffectComponent, StatusEffectAppliedEvent>(OnStatusEffectApplied);
```
So that it's not trying to run the event handler at all unless it's an event we care about here. That said, this _is_ more than a third of the status effect components that we have today, so I'm not sure how the breakdown works out.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Gargantua.cs:
```diff
@@ -0,0 +1,691 @@
+            return;
+        }
+        
+        var gargantua = EnsureComp<GargantuaComponent>(uid);
```
Do we really want to be adding the Gargantua component to the entity here, vs a TryComp() || return?

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/VampireSystem.Abilities.cs:
```diff
@@ -0,0 +1,726 @@
+        foreach (var ent in _lookup.GetEntitiesIntersecting(_transform.ToMapCoordinates(coords), LookupFlags.Static))
+        {
+            // Skip non anchored entities
+            if (!TryComp(ent, out TransformComponent? entTransform) || !entTransform.Anchored)
```
```suggestion
            if (!Transform(ent).Anchored)
```

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/VampireSystem.Abilities.cs:
```diff
@@ -0,0 +1,726 @@
+
+            UpdateFullPower(uid, comp);
+
+            var biteSound = new SoundPathSpecifier("/Audio/Effects/bite.ogg");
```
Move to readonly class member instead of making a new path specifier each bite.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/VampireSystem.Abilities.cs:
```diff
@@ -0,0 +1,726 @@
+            // Threshold bonuses at 300 TotalBlood
+            if (comp.TotalBlood >= 300)
+            {
+                switch (comp.ChosenClass)
```
If you do swap the vampire classes to being their own components, this probably becomes a VampireBloodThresholdBonusEvent and associated handlers?

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/VampireSystem.Umbrae.cs:
```diff
@@ -0,0 +1,670 @@
+        args.Handled = true;
+    }
+
+    private void OnShadowSnareStepAttempt(EntityUid uid, ShadowSnareComponent component, ref StepTriggerAttemptEvent args) => args.Continue = true;
```
These ShadowSnareComponent event handlers can probably be moved to a ShadowSnareSystem.

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/Components/SanguinePoolComponent.cs:
```diff
@@ -0,0 +1,48 @@
+    [DataField]
+    public SoundSpecifier ExitSound = new SoundPathSpecifier("/Audio/_Starlight/Effects/vampire/exit_blood.ogg");
+
+    [ViewVariables]
```
If it serializes, VVs, and generally works nicely, this might be better as
```cs
public (EntityUid Grid, Vector2i Tile)? LastTrail;
```

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/Components/SanguinePoolComponent.cs:
```diff
@@ -0,0 +1,48 @@
+    ///     How often to spawn a trailing puddle while gliding around.
+    /// </summary>
+    [DataField]
+    public float TrailInterval = 0.3f;
```
This gets set here and in VampireSystem.Hemomancer.cs, but I don't see it getting read anywhere?

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/Components/VampireComponent.cs:
```diff
@@ -0,0 +1,195 @@
+    public float HolyPlaceRange = 4f;
+
+    [DataField]
+    public string HolyWaterReagentId = "Holywater";
```
`ProtoId<ReagentPrototype>`?

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/Components/VampireSunlightComponent.cs:
```diff
@@ -0,0 +1,80 @@
+    ///     Genetic damage applied each tick when the vampire has no blood
+    /// </summary>
+    [DataField]
+    public float GeneticDamagePerInterval = 10f;
```
Health damage is computed in `FixedPoint2`s and should be referred to in components via `FixedPoint2`s or `DamageSpecifier`s.

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/Components/VampireSunlightComponent.cs:
```diff
@@ -0,0 +1,80 @@
+    ///     How long a vampire can linger in space before they start taking damage
+    /// </summary>
+    [DataField]
+    public float GracePeriod = 1.5f;
```
Please use `TimeSpan`s for these spans of time.

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/VampireEvents.cs:
```diff
@@ -0,0 +1,553 @@
+    /// How many seconds do we need to mute entity in front of glare source.
+    /// </summary>
+    [DataField]
+    public int MuteDuration = 8;
```
Can this be a `TimeSpan`?

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/VampireEvents.cs:
```diff
@@ -0,0 +1,553 @@
+    public EntProtoId TendrilsPuddlePrototype = "PuddleBlood";
+
+    [DataField]
+    public float Delay = 1.0f;
```
More `TimeSpan`s.

Penguinwizzard reviewed (changes requested):
General themes:
1. The Vampire subclasses are IMO too tightly wound into VampireSystem. It would be nice if adding a new subclass just meant a new Component for the class (which could inherit from an abstract VampireClassComponent or something), a System for that component, and a VampireClassPrototype instance in the yaml which specifies the name, the component to add, and any data needed for the selection menu.
2. There's a few vampire abilities that basically put you into a state that has some effects. IMO these should probably be handled by adding a component to the user, rather than keeping a bool tracking if you're in that state or not.
3. For the Component/Event definitions, please track damage amounts using a FixedPoint2, and consider specifying the damage outright in the component using a DamageSpecifier. Similarly, durations/times/intervals should be TimeSpans.

CringeCursed committed:
Apply Suggestions + add ability to devour mouse(actually anything with VampireDevour component) to restore VampireFed

CringeCursed committed:
erm...

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/Classes/DantalionSystem.cs:
```diff
@@ -4,31 +4,61 @@
     {
+        base.Initialize();
+
         SubscribeLocalEvent<VampireComponent, VampireEnthrallActionEvent>(OnEnthrall);
```
Would it be possible to change most/all of the event listeners in this file to look for events on entities with DantalionComponent, rather than entities with VampireComponent? That'd make the events a little more targeted, and let you avoid having the HasComp<Dantalion> at the start of almost every method here (although you'd need a TryComp<Vampire>).

CringeCursed replied:
Sure, changed it up to do so.

CringeCursed committed:
Refactor vampire devour event names for consistency

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/Classes/DantalionSystem.cs:
```diff
@@ -241,13 +279,39 @@ private int GetThrallLimit(VampireComponent comp, DantalionComponent dantalion)
-#endregion
+    private void HealDantalionThralls(EntityUid uid)
+    {
+        if (!TryComp<DantalionComponent>(uid, out var dantalion) || dantalion.Thralls.Count == 0)
```
In the one place you call this you already have the DantalionComponent reference; changing the argument to an `Entity<DantalionComponent>` would let you avoid this `TryComp`.

CringeCursed replied:
Changed, ty.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/Classes/DantalionSystem.cs:
```diff
@@ -241,13 +279,39 @@ private int GetThrallLimit(VampireComponent comp, DantalionComponent dantalion)
-    public void OnPacify(EntityUid uid, VampireComponent comp, ref VampirePacifyActionEvent args)
+        foreach (var thrall in dantalion.Thralls.ToArray())
+        {
+            if (!Exists(thrall))
```
You have a few cases in this file where you enumerate through `dantalion.Thralls` and do the `!Exists` and `Master != uid` checks - would it make sense to have a method like
```cs
private IEnumerable IterateAndCheckThralls(Entity<DantalionComponent> dantalion)
{
  foreach (var thrall in dantalion.Thralls.ToArray())
  {
    if (
        !Exists(thrall)
        || !TryComp<VampireThrallComponent>(thrall, out var thrallComp)
        || thrallComp.Master != dantalion.Owner
    )
    {
      dantalion.Thralls.Remove(thrall);
      continue;
    }

    yield return thrall;
  }
}
```
and call that for any cases with thrall loops?

CringeCursed replied:
Yep, maked an general helper, appreciate you.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/Classes/DantalionSystem.cs:
```diff
@@ -625,7 +680,7 @@ private void ApplyBloodBondDamageSharing(EntityUid vampire, List<EntityUid> thra
                 continue;
 
             var totalDamage = damageable.TotalDamage.Float();
```
Is `TotalDamage` what you want to look at here? Right now, if a vampire has a ton of genetic damage, this would show them as having damage, but it wouldn't get healed by the brute/burn heals later in this method, meaning that the other participants would take brute for no gain.

CringeCursed replied:
Damn a good point. Tweaked it so it only takes only "healable" damage.

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/ShadowSnareSystem.cs:
```diff
@@ -0,0 +1,91 @@
+            return;
+
+        // Apply brute damage
+        var damage = new DamageSpecifier();
```
This DamageSpecifier should _probably_ be static on the class or a member of the component.

CringeCursed replied:
moved to component

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.Abilities.cs:
```diff
@@ -109,7 +175,21 @@ private void ApplyDamage(EntityUid target, string damageTypeId, float amount, En
+        _damageableSystem.TryChangeDamage(target, spec, true, origin: origin);
+    }
+
+    internal void ApplyDamage(EntityUid target, string damageTypeId, FixedPoint2 amount, EntityUid? origin = null)
```
I'm not a huge fan of the ApplyDamage/ApplyHealing functions here. Most of their callsites should be providing a DamageSpecifier straight to _damageableSystem.TryChangeDamage, which would avoid the need for all of these string maps and lookups, not to mention the multiple calls to TryChangeDamage if you want to e.g. heal three damage types at once, like the thrall healing.

CringeCursed replied:
Removed Helpers and changed it to direct TryChangeDamage with _composite_ DamageSpecifier

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.cs:
```diff
@@ -173,18 +169,24 @@ private void HandleSpaceExposure(EntityUid uid, VampireComponent vampire, Vampir
-        sunlight.DamageAccumulator += frameTime;
-        var interval = Math.Max(sunlight.DamageInterval, 0.1f);
+
+        sunlight.DamageAccumulator += frameSpan;
```
Can we define DamageAccumulator as simply TimeInSpace - GracePeriod?

CringeCursed replied:
Sure, made it up to do so

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.cs:
```diff
@@ -173,18 +169,24 @@ private void HandleSpaceExposure(EntityUid uid, VampireComponent vampire, Vampir
         }
 
-        sunlight.TimeInSpace += frameTime;
+        var frameSpan = TimeSpan.FromSeconds(frameTime);
```
Rather than tracking TimeInSpace by accumulating frameTime, could we record TimeEnteredSpace, and use AutoGenerateComponentPause and such to handle updating it when the entity is paused/unpaused? That'll result in modifying the component less frequently.

CringeCursed replied:
We now store TimeEnteredSpace and schedule future timestamps (NextDamageTime, NextWarningPopup) instead of mutating an accumulator every update. The fields in SunlightComponent are pause  aware via AutoGenerateComponentPause + AutoPausedField

Penguinwizzard commented on Content.Server/_Starlight/Antags/Vampires/Systems/VampireSystem.cs:
```diff
@@ -444,35 +445,21 @@ private void TryRefreshVampireAction(EntityUid owner, EntityUid? actionEntity)
         _actions.SetEnabled(action.AsNullable(), enabled);
     }
 
     private void TryGrantClassAbilities(EntityUid uid, VampireComponent comp)
```
Out of curiosity, why do we need to continually grant abilities in the update loop, as opposed to when the class component is added? To handle polymorphs?

CringeCursed replied:
Kinda, but this way of granting abilities in update was before Polymorph ability, my concern was that polymorph can break "grant abilties"
But i switched to event driven pattern.

Penguinwizzard commented on Content.Server/_Starlight/GameTicking/Rules/VampireRuleSystem.cs:
```diff
@@ -42,20 +44,18 @@ public bool MakeVampire(EntityUid target, VampireRuleComponent rule)
+
+        _role.MindHasRole<VampireRoleComponent>(mindId, out var vampRole);
+        _role.MindHasRole<RoleBriefingComponent>(mindId, out var briefingComp);
+        if (vampRole is not null && briefingComp is null)
```
```suggestion
        if (
             _role.MindHasRole<VampireRoleComponent>(mindId, out var vampRole)
          && _role.MindHasRole<RoleBriefingComponent>(mindId, out var briefingComp)
        )
```

CringeCursed replied:
ty

Penguinwizzard commented on Content.Shared/_Starlight/Antags/Vampires/Components/VampireComponent.cs:
```diff
@@ -155,10 +103,10 @@ public sealed partial class VampireComponent : Component
-    public string HolyWaterReagentId = "Holywater";
+    public ProtoId<ReagentPrototype> HolyWaterReagentId = "Holywater";
 
     public TimeSpan NextHolyWaterTick = TimeSpan.Zero;
```
This may warrant use of AutoGenerateComponentPause.

CringeCursed committed:
move test to sl folder

Penguinwizzard reviewed (changes requested):
Some more generally smaller changes. I really like the edits from the last round of updates you made!

DocMoth commented:
holding up my hands and sharing my energy with ya right now. vamps soon

CringeCursed committed:
Refactor vampire damage and healing mechanics

- Updated GargantuaSystem to use DamageSpecifier for applying healing.
- Refactored HemomancerSystem to utilize DamageSpecifier for damage application and healing.
- Modified UmbraeSystem to apply damage using DamageSpecifier.
- Simplified ShadowSnareSystem to directly use DamageSpecifier for damage application.
- Removed redundant damage type caching from VampireSystem and replaced with direct DamageSpecifier usage.
- Enhanced vampire healing logic to utilize DamageSpecifier for various healing types.
- Introduced new starvation mechanics in VampireComponent, affecting movement speed when blood fed is empty.
- Created SharedVampireStarvationSystem to handle movement speed modifications based on blood fullness.
- Updated vampire events to include VampireProgressionChangedEvent for tracking blood-related changes.
im tired boss :(

CringeCursed committed:
apply suggestio

CringeCursed committed:
one more

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
fix linter

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
Update  collective_mind.yml

Removed 'Resomi' and 'Avali' from requiredTags and added 'Nexus'.

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
Fix blood solution

CringeCursed committed:
Merge branch 'Starlight-dev' into vamps-rework-continuew

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
some fixes

CringeCursed committed:
fix blood_pool animation

CringeCursed committed:
some chaplain buffs, loc. fixes

CringeCursed committed:
chaplain part 2

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
few fixes

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew

CringeCursed committed:
123

CringeCursed committed:
upd rt

CringeCursed committed:
Merge branch 'starlight-dev' into vamps-rework-continuew
